### PR TITLE
Slide when diagonal movement is blocked

### LINKED ISSUE #900: Running along walls and then pressing to run into the wall as well, causes you to stop
# Preliminary information
Running along walls and then pressing to run into the wall as well, causes you to stop 

## Category
 this is a  BUG,

## Bug: Release or Development?
it was  on the Latest steam version Beta 5 

# Report
## Current Behavior
Running along walls and then pressing to run into the wall as well, causes you to stop 

## Expected/Wanted/Requested Behavior
like in TG code you keep on running along the wall instead of stopping

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Find long stretch of wall
2.Run along 
3.Press key to go into the wall while still pressing the same key to go along the wall

## OS used:
Windows

## Network
server That is default on steam

### PR DESCRIPTION
**Pull request by chairbender**
### Purpose
Fixes #900. Refactors a few things related to how bumps are checked.

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer

### Notes:
I'll add a PR comment in PlayerSync.Client.cs to show the actual line that was changed (Ctrl+K,Ctrl+D strikes again).

See the method TrySlide in PlayerSync.cs to see the logic for sliding.

### CODE REVIEW

File: UnityProject/Assets/Scripts/Player/PlayerSync.Client.cs [hunk 1]
```diff
@@ -5,423 +5,495 @@
...
+		{
+			//				Logger.LogTraceFormat( "{0} requesting {1} ({2} in queue)", Category.Movement, gameObject.name, action.Direction(), pendingActions.Count );
 
-			yield return YieldHelper.DeciSecond;
-			MoveCooldown = false;
-		}
+			if (isGrounded && playerState.Active)
+			{
+				//RequestMoveMessage.Send(action);
+				// Fix for #900
```
Comment (chairbender) [inline UnityProject/Assets/Scripts/Player/PlayerSync.Client.cs:97]: The actual change for this file is just these few lines

File: UnityProject/Assets/Scripts/Objects/PushPull.cs [hunk 2]
```diff
@@ -438,6 +461,7 @@ public bool TryPush( Vector2Int dir, float speed = Single.NaN )
 	}
 
 	public bool TryPredictivePush( Vector3Int from, Vector2Int dir, float speed = Single.NaN ) {
+		//how does this differ from CanPush
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Objects/PushPull.cs]: this one has clientside checks, and one that you extracted to `CanPush` has serverside checks

**Review by fomalsd — CHANGES_REQUESTED** [review]
The code looks good, it feels really nice on server, but when testing with high ping client I've got a lot of lerpbacks when passing a door:
new, with sliding:
gif0
old, without sliding:
gif1

refer to [this guide](https://github.com/unitystation/unitystation/wiki/Building-And-Testing#set-up-ping-simulation) to set up ping simulation

looking at how move numbers get reset, it means that server doesn't approve some of the movements towards doors

**Comment by chairbender** [comment]
Thanks @fomalsd , good catch. I did test with high ping but didn't notice the difference that only occurred at the doors. Looks like you have some sort of feature turned on which makes it easier to see the positions (the green and white boxes), I'll inquire about that on Discord.

Should be fixed now.

File: UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs [hunk 3]
```diff
@@ -395,18 +400,29 @@ private PlayerState NextStateServer(PlayerState state, PlayerAction action)
...
+	/// <returns></returns>
 	private IEnumerator TryInteract( Vector3 currentPosition, Vector3 direction ) {
 		var worldPos = Vector3Int.RoundToInt(currentPosition);
 		var worldTarget = Vector3Int.RoundToInt(currentPosition + direction);
 
 		InteractDoor(worldPos, worldTarget);
 
 //		Logger.LogTraceFormat( "{0} Interacting {1}->{2}, server={3}", Category.Movement, Time.unscaledTime*1000, worldPos, worldTarget, isServer );
-		InteractPushable( worldTarget, direction );
+		InteractPushable(worldPos, direction );
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs:425]: Why did you change this interaction behaviour btw? was there a bug of some sort before related to this?
I think it introduced this side effect of being unable to get out of closed doors by bumping using movement:

before
scr 2019-01-11 before
after
scr 2019-01-11 after

**Review by fomalsd — COMMENTED** [review]
Yeah, server no longer rollbacks player for walking through doors, really nice!
Still there's that one small thing with modified bump origin I'm not quite sure about (not that players are supposed to get stuck in doors in the first place though)

**Review by fomalsd — APPROVED** [review]

File: UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs (hunk 3 above)
Comment (chairbender) [reply UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs:425]: That solved a bug I was encountering when pushing. I can add logic to ensure the previous behavior works.

Comment (chairbender) [reply UnityProject/Assets/Scripts/Player/PlayerSync.Server.cs:425]: NVM, this will be fixed as part of a different ticket #1110 so it would be throwaway work to fix it here since the way door closing works will be different

**Comment by chairbender** [comment]
ready to merge
